group-coordinator/src/main/java/org/apache/kafka/coordinator/group/RecordHelpers.java

```diff
@@ -357,6 +366,94 @@ public static Record newCurrentAssignmentTombstoneRecord(
         );
     }
 
+    /**
+     * Creates a GroupMetadata record.
+     *
+     * @param group              The generic group.
+     * @param memberAssignments  The assignment by member id.
+     * @param metadataVersion    The metadata version.
+     * @return The record.
+     */
+    public static Record newGroupMetadataRecord(
+        GenericGroup group,
+        Map<String, byte[]> memberAssignments,
```

Comment:
the assignment comes directly from the sync group request, i don't think we need to store the assignment inside the group

Reply:
hmm... don't we store it in MemberMetadata in the current implementation? We set it in `setAndPropagateAssignment`. I think that we need it because we need the ability to provide it the the member at any time.

Reply:
i think you're referring to https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala#L623

the in-memory state is updated after we successfully append to the log which so we won't have this information when we generate the records to append, no?

Reply:
yeah, that's right. the generic group is a bit weird from this regard because it does not use timeline data structures internally so we could just update it as well (like we do for subscriptions for instance). the part which is weird here is that we take everything from the group but the assignment.

Reply:
i think you're saying that ideally we want to create a record solely from the group i.e. update the group with the assignment and pass it into this method. However, we can't because we don't use timeline data structures that could revert if append fails. is my understanding correct?

Reply:
it is actually the other way around, the generic group is mutable so we could mutate it whenever we want here, i think.